packages/gatsby/src/utils/source-nodes.js

```diff
@@ -65,29 +65,51 @@ function getStaleNodes(state, nodes) {
   })
 }
 
+let firstRun = true
+
 /**
  * Find all stale nodes and delete them
  */
 function deleteStaleNodes(state, nodes) {
   const staleNodes = getStaleNodes(state, nodes)
 
   if (staleNodes.length > 0) {
-    staleNodes.forEach(node => deleteNode({ node }))
+    staleNodes.forEach(node => {
+      if (!firstRun && node.internal.type === `SitePage`) {
+        // This is a HACK!
+        // SitePage nodes are created in stateful way
+        // but we don't have sourceNodesStatefully to support
+        // it properly. So instead we avoid deleting them
+        // on non-first calls.
+        return
+      }
+      deleteNode({ node })
+    })
```

Comment:
Just wanted to make this change visible - as mentioned - it's hack and to un-hack it we will need to implement https://github.com/gatsbyjs/gatsby/issues/4682 (might do this as part of this PR, as I'm not overly enthusiastic about this hack)